### PR TITLE
Add a channel_axis argument to functions in the skimage.color module

### PR DESCRIPTION
**Pull request by grlee77**
## Description

This PR adds a new `channel_axis` argument to color conversion functions as discussed a bit in #5439 and corresponding recent API-related meetings.

For most functions this is pretty simple, often just requiring adding the `channel_axis` keyword and the pre-existing `channel_as_last_axis` decorator. Most of the work was in modifying the test cases to test each function with additional channel axis choices.

One specific case that is a bit tricky is how to handle the `adapt_rgb` decorator so that it supports channels along any axis instead of always assuming the last axis. I implemented something for that in commit 8063049 that involves inspection of keyword arguments of the decorated function, but I think this is possibly too magical and is perhaps not worth it.  It doesn't help that `adapt_rgb` is also one place where we apparently still try to guess if an image is RGB/RGBA based on its shape. The only function in the library that uses `adapt_rgb` is `skimage.exposure.equalize_adapthist`. If we don't want this change I can just revert that commit and/or leave it to a potential future follow-up PR.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/color/colorlabel.py [hunk 1]
```diff
@@ -108,20 +108,30 @@ def label2rgb(label, image=None, colors=None, alpha=0.3,
...
+        to channels. If `image` is provided, this must also match the axis of
+        `image` that corresponds to channels.
+
+        .. versionadded:: 0.19
+           ``channel_axis`` was added in 0.19.
 
     Returns
     -------
-    result : array of float, shape (M, N, 3)
+    result : array of float, shape (..., 3, ...)
```
Comment (mkcor) [inline skimage/color/colorlabel.py:121]: Maybe update line 84 along the same lines: `image : array, shape (M, N) or (..., 3, ...), optional`?

**Comment by mkcor** [comment]
> One specific case that is a bit tricky is how to handle the `adapt_rgb` decorator so that it supports channels along any axis instead of always assuming the last axis.

I would be very tempted to drop `skimage/color/adapt_rgb.py` altogether.

> It doesn't help that `adapt_rgb` is also one place where we apparently still try to guess if an image is RGB/RGBA based on its shape.

Exactly. It's not robust at all and it doesn't generalize to nD. I'm a little scared when I read: "A very simple check of the array shape is used for detecting RGB images, so `adapt_rgb` is not recommended for functions that support 3D volumes or color images in non-RGB spaces." (note at the bottom of https://scikit-image.org/docs/stable/auto_examples/color_exposure/plot_adapt_rgb.html where we document/encourage the use of the `adapt_rgb` decorator) :scream: 

> The only function in the library that uses `adapt_rgb` is `skimage.exposure.equalize_adapthist`.

I searched all GitHub and found that this [project](https://github.com/lauraluebbert/epi_image_analysis) by @lauraluebbert uses `skimage.exposure.equalize_adapthist`. I'll double check when Binder starts, but it looks like the function is used on single-channel inputs (even though the analyzed dataset consists of colour images).

**Comment by grlee77** [comment]
Yeah I think it is best if I just rebase without commit 3d82e2c or 8063049 (I think all `adaptrgb`-related changes were in those two commits). We can decide about potentially dropping the decorator altogether elsewhere.

**Comment by grlee77** [comment]
I decided to use `git revert` instead of `git rebase -i` so the commits are still in the history

File: skimage/color/colorlabel.py (hunk 1 above)
Comment (grlee77) [reply skimage/color/colorlabel.py:121]: done

File: skimage/color/adapt_rgb.py [hunk 2]
```diff
@@ -9,14 +9,14 @@
 __all__ = ['adapt_rgb', 'hsv_value', 'each_channel']
 
 
-def is_rgb_like(image):
+def is_rgb_like(image, channel_axis=-1):
```
Comment (mkcor) [inline skimage/color/adapt_rgb.py:12]: Ok for this one, it doesn't really hurt! :sweat_smile:

File: skimage/_shared/utils.py [hunk 3]
```diff
@@ -457,6 +455,41 @@ def slice_at_axis(sl, axis):
...
+    >>> arr = rng.random(7)
+    >>> _reshape_nd(arr, 2, 0).shape
+    (7, 1)
+    >>> _reshape_nd(arr, 3, 1).shape
+    (1, 7, 1)
+    >>> _reshape_nd(arr, 4, -1).shape
+    (1, 1, 1, 7)
+    """
+    if arr.ndim != 1:
+        raise ValueError("arr must be a 1d array")
```
Comment (mkcor) [inline skimage/_shared/utils.py:487]:
```suggestion
        raise ValueError("arr must be a 1D array")
```

File: skimage/color/colorconv.py [hunk 4]
```diff
@@ -112,42 +124,60 @@ def convert_colorspace(arr, fromspace, tospace):
...
     -------
-    out : (..., 3) ndarray
+    out : (..., 3, ...) ndarray
         The image in RGB format. Same dimensions as input.
 
     Raises
     ------
     ValueError
-        If `rgba` is not at least 2-D with shape (..., 4).
+        If `rgba` is not at least 2-D with shape (..., 4, ...).
```
Comment (mkcor) [inline skimage/color/colorconv.py:180]:
```suggestion
        If `rgba` is not at least 2D with shape (..., 4, ...).
```

File: skimage/color/colorconv.py [hunk 5]
```diff
@@ -112,42 +124,60 @@ def convert_colorspace(arr, fromspace, tospace):
...
 
     return dtype.img_as_float(arr, force_copy=force_copy)
 
 
-def rgba2rgb(rgba, background=(1, 1, 1)):
+def _validate_channel_axis(channel_axis, ndim):
+    if not isinstance(channel_axis, int):
+        raise TypeError("channel_axis must be an integer")
+    if channel_axis < -ndim or channel_axis >= ndim:
+        raise ValueError("channel_axis exceeds array dimensions")
```
Comment (mkcor) [inline skimage/color/colorconv.py:151]: Did you consider raising numpy's `AxisError` in this case?

**Review by mkcor — COMMENTED** [review]
I'm only half-way through, but it's getting late; will finish during the week!

File: skimage/color/colorconv.py [hunk 6]
```diff
@@ -114,18 +126,22 @@ def convert_colorspace(arr, fromspace, tospace):
...
     """Check the shape of the array and convert it to
     floating point representation.
     """
     arr = np.asanyarray(arr)
 
-    if arr.shape[-1] != 3:
-        raise ValueError("Input array must have a shape == (..., 3)), "
-                         f"got {arr.shape}")
+    if arr.shape[channel_axis] != 3:
+        msg = ("the input array must have shape 3 along `channel_axis`, "
```
Comment (mkcor) [inline skimage/color/colorconv.py:142]:
```suggestion
        msg = (f"the input array must have size 3 along `channel_axis`, "
```
as in https://github.com/scikit-image/scikit-image/pull/5462/files#diff-3b92ef8526333aa6b6c52d8971304d93c4470d4ac47591fefe703145a43d6277R54?
shape would be like (3,) :wink:

File: skimage/color/colorconv.py [hunk 7]
```diff
@@ -168,9 +198,11 @@ def rgba2rgb(rgba, background=(1, 1, 1)):
     >>> img_rgb = color.rgba2rgb(img_rgba)
     """
     arr = np.asanyarray(rgba)
+    _validate_channel_axis(channel_axis, arr.ndim)
+    channel_axis = channel_axis % arr.ndim
 
-    if arr.shape[-1] != 4:
-        msg = ("the input array must have shape == (..., 4)), "
+    if arr.shape[channel_axis] != 4:
+        msg = ("the input array must have shape 4 along `channel_axis`, "
```
Comment (mkcor) [inline skimage/color/colorconv.py:205]:
```suggestion
        msg = (f"the input array must have size 4 along `channel_axis`, "
```
(or length)

**Review by mkcor — COMMENTED** [review]
I'm confused: I couldn't find `channel_as_last_axis` defined in `skimage/_shared/utils.py`--where is it?

File: skimage/color/colorconv.py (hunk 5 above)
Comment (grlee77) [reply skimage/color/colorconv.py:158]: yeah, I probably should have done that. I will update it

File: skimage/color/colorconv.py [hunk 8]
```diff
@@ -135,26 +151,40 @@ def _prepare_colorarray(arr, force_copy=False):
     return _func(arr, force_copy=force_copy)
 
 
-def rgba2rgb(rgba, background=(1, 1, 1)):
+def _validate_channel_axis(channel_axis, ndim):
+    if not isinstance(channel_axis, int):
+        raise np.AxisError("channel_axis must be an integer")
```
Comment (mkcor) [inline skimage/color/colorconv.py:156]: @grlee77 it would be more consistent if this one were a `TypeError`, from what I tested out for #5420 (see line https://github.com/scikit-image/scikit-image/pull/5420/files#diff-a62ddc4b2d437b0b4ee3a01b2eba9a2cb7a148f9cf03f1e6902e6307a9596e2fR61).

Comment (grlee77) [reply skimage/color/colorconv.py]: Agreed. I have switched this one to `TypeError`

**Comment by mkcor** [comment]
> I'm confused: I couldn't find `channel_as_last_axis` defined in `skimage/_shared/utils.py`--where is it?

False alarm, sorry, I don't know what happened to my sed/grep... :confused: 

> the pre-existing `channel_as_last_axis` decorator

line 220 of expected file in current `main` :sweat_smile:

**Review by mkcor — APPROVED** [review]

File: skimage/color/colorconv.py [hunk 9]
```diff
@@ -852,77 +940,36 @@ def gray2rgba(image, alpha=None):
...
+    if np.isscalar(alpha):
+        alpha = np.full(arr.shape, alpha, dtype=arr.dtype)
+    elif alpha.shape != arr.shape:
+        raise ValueError("alpha.shape must match image.shape")
+    rgba = np.stack((arr,) * 3 + (alpha,), axis=channel_axis)
     return rgba
 
 
-def gray2rgb(image, alpha=None):
+def gray2rgb(image, *, channel_axis=-1):
```
Comment (rfezzani) [inline skimage/color/colorconv.py:951]: The `alpha` argument may had be removed in a different PR, but it may be more complicated. Please remove the corresponding item in TODO.txt :wink:

File: skimage/color/tests/test_colorconv.py [hunk 10]
```diff
@@ -77,9 +77,14 @@ class TestColorconv():
                           ])
 
     # RGBA to RGB
-    def test_rgba2rgb_conversion(self):
+    @pytest.mark.parametrize("channel_axis", [0, 1, 2, -1])
```
Comment (rfezzani) [inline skimage/color/tests/test_colorconv.py:80]: In the following tests, may be can we test all the negative axis value range (not only -1)
```suggestion
    @pytest.mark.parametrize("channel_axis", [0, 1, 2, -1, -2, -3])
```

**Review by rfezzani — COMMENTED** [review]
Thank you @grlee77, I left some suggestions, otherwise it LGTM :wink:

File: skimage/color/colorconv.py (hunk 9 above)
Comment (grlee77) [reply skimage/color/colorconv.py:951]: Apparently I did already remove `alpha` and the TODO item as part of #5463 so perhaps we should merge that first to make this cleaner.

**Comment by grlee77** [comment]
Thanks @rfezzani, I implemented the requested changes in the last few commits.

**Comment by rfezzani** [comment]
@grlee77, if you prefer to merge #5463 prior to this one, there is a conflict in `colorlabel.py` there :confused:

**Review by rfezzani — APPROVED** [review]

**Comment by rfezzani** [comment]
@grlee77, we will be ready for merge after conflict resolution :wink:

**Comment by rfezzani** [comment]
:tada: Thank you again @grlee77
